docs/source/design/credentials.rst

```diff
@@ -28,11 +33,60 @@ These locations are tried in the following order, and the first successful looku
+
+Providers
+=========
+
+Providers are associating credentials with a context for using them and are defined by configuration files.
+A single provider is represented by :class:`Provider` object and the list of available providers is represented by the :class:`Providers` class.
+A provider is identified by a label and stored in a dedicated config file per provider named `LABEL.cfg`.
+Such a file can reside in a dataset (under `.datalad/providers/`), at the user level (under `{user_config_dir}/providers`), at the system level (under `{site_config_dir}/providers`) or come packaged with the datalad distribution (in directory `configs` next to `providers.py`).
+Such a provider specifies a regular expression to match URLs against and assigns authenticator abd credentials to be used for a match.
+Credentials are referenced by their label, which in turn is the name of another section in such a file specifying the type of the credential.
+References to credential and authenticator types are strings that are mapped to classes by the following dict definitions:
+
+- ``datalad.downloaders.AUTHENTICATION_TYPES``
+- ``datalad.downloaders.CREDENTIAL_TYPES``
+
+Available providers can be loaded by ``Providers.from_config_files`` and ``Providers.get_provider(url)`` will match a given URL against them and return the appropriate `Provider` instance.
+A :class:`Provider` object will determine a downloader to use (derived from :class:`BaseDownloader`), based on the URL's protocol.
+
+Note, that the provider config files are not currently following datalad's general config approach.
+Instead they are special config files, read by :class:`configparser.ConfigParser` that are not compatible with `git-config` and hence the :class:`ConfigManager`.
+
+There are currently two ways of storing a provider and thus creating its config file: ``Providers.enter_new`` and ``Providers._store_new``.
+The former will only work interactively and provide the user with options to choose from, while the latter is non-interactive and can therefore only be used, when all properties of the provider config are known and passed to it.
+There's no way at the moment to store an existing :class:`Provider` object directly.
+
+Integration with Git
+====================
+
+In addition, there's a special case for interfacing `git-credential`: A dedicated :class:`GitCredential` class is used to talk to Git's ``git-credential`` command instead of the keyring wrapper.
+This class has identical fields to the :class:`UserPassword` class and thus can be used by the same authenticators.
+Since Git's way to deal with credentials doesn't involve labels but only matching URLs, it is - in some sense - the equivalent of datalad's provider layer.
+However, providers don't talk to a backend, credentials do.
+Hence, a more seemless integration requires some changes in the design of datalad's credential system as a whole.
```

Comment:
seemless -> seamless

I would also recommend saying that it /would/ require changed to the design. Make it clear that it's hypothetical.